### PR TITLE
Code refactor of ConstrainedOptimizer

### PR DESCRIPTION
**Pull request by gallego-posada**
- Code refactor in the ConstrainedOptimizer class
- Multipliers and ExtraOptimizers have been separated from the logic in ConstrainedOptimizer
- Fixes bugs in the implementation of the computation of the Lagrangian and the backward propagation thereof
- Includes significant improvements in documentation
- Implements dual_resets
- Includes test file with toy classification task
- Enables passing primal and dual optimizer kwargs

### CODE REVIEW

File: src/torch_constrained/constrained_optimizer.py [hunk 1]
```diff
@@ -1,359 +1,289 @@
...
-            lr_y,
+            primal_optim_class: Type[torch.optim.Optimizer],
+            primal_kwargs: Dict,
             primal_parameters,
-            augmented_lagrangian_coefficient=False,
+            dual_optim_class: Optional[Type[torch.optim.Optimizer]]=None,
+            dual_kwargs: Optional[Dict]=None,
+            ineq_init: Optional[List[torch.Tensor]]=None,
+            eq_init: Optional[List[torch.Tensor]]=None,
+            augmented_lagrangian_coeff=False,
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:19]: in general, modifying the variable names with no explicit reason is not too nice/useful, and increases the review load, can you revert it?

Comment (manuel-delverme) [reply src/torch_constrained/constrained_optimizer.py:19]: Unless there was a reason for the change

File: src/torch_constrained/constrained_optimizer.py [hunk 2]
```diff
@@ -1,359 +1,289 @@
...
             alternating=False,
-            shrinkage: Optional[Callable] = None,
-            dual_dtype=None,
-            primal_kwargs={},
-            dual_kwargs={},
-    ):
+            shrinkage: Optional[Callable]=None,
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:21]: https://pep8.org/#whitespace-in-expressions-and-statements

File: src/torch_constrained/constrained_optimizer.py [hunk 3]
```diff
@@ -1,359 +1,289 @@
...
+            dual_reset=False,
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:22]: something like, dual_reset_on_feasibility based on the comment later, could make a more explicit name.
is it for ineq? for eq only?

File: src/torch_constrained/constrained_optimizer.py [hunk 4]
```diff
@@ -1,359 +1,289 @@
...
-        return self.state["inequality_multipliers"]
+        Args:
+            eq_defect: Defects for equality constraints 
+            ineq_defect: Defects for inequality constraints.
+        """
 
-    @property
-    def equality_multipliers(self):
-        return self.state["equality_multipliers"]
+        if self.verbose: print('Initializing dual variables')
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:225]: this should be using the logger util to avoid interfering with the user code.

File: src/torch_constrained/constrained_optimizer.py [hunk 5]
```diff
@@ -1,359 +1,289 @@
...
-        self.primal_optimizer = loss_optimizer(primal_parameters, lr_x, **primal_kwargs)
-        self.dual_optimizer_class = functools.partial(constraint_optimizer, lr=lr_y, **dual_kwargs)
-
-        self.dual_optimizer = None
-        self.augmented_lagrangian_coefficient = augmented_lagrangian_coefficient
-        self.alternating = alternating
-        self.shrinkage = shrinkage
-        self.dual_dtype = dual_dtype
+        self.primal_optimizer = primal_optim_class(primal_parameters, **primal_kwargs)
+        self.primal_parameters = primal_parameters
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:30]: why do we need to store the parameters?

File: src/torch_constrained/constrained_optimizer.py [hunk 6]
```diff
@@ -1,359 +1,289 @@
...
+
+        # Initialization values for the dual variables
+        self.dual_init_done = False
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:33]: this could be a private variable

File: src/torch_constrained/constrained_optimizer.py [hunk 7]
```diff
@@ -1,359 +1,289 @@
...
+        self.ineq_init = ineq_init
+        self.eq_init = eq_init
+
+        # Create flag for execution in un-constrained setting
+        self.is_constrained = dual_optim_class is not None
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:38]: the problem could be constrained, but without a dual optimizer (regularized), `self.dual_optimizer is None` or something similar should be enough.

File: src/torch_constrained/constrained_optimizer.py [hunk 8]
```diff
@@ -1,359 +1,289 @@
...
+
+        # Initialization values for the dual variables
+        self.dual_init_done = False
+        self.ineq_init = ineq_init
+        self.eq_init = eq_init
+
+        # Create flag for execution in un-constrained setting
+        self.is_constrained = dual_optim_class is not None 
+
+        if self.is_constrained:
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:40]: `if dual_optim_class is not None `

File: src/torch_constrained/constrained_optimizer.py [hunk 9]
```diff
@@ -1,359 +1,289 @@
...
+            # Important: the dual optimizer is instantiated in 'init_dual_variables'.
+            self.dual_optimizer_class = partial(dual_optim_class, **dual_kwargs)
+            self.dual_reset = dual_reset
+
+            # Other optimization and lagrangian options
+            self.augmented_lagrangian_coeff = augmented_lagrangian_coeff
+            self.alternating = alternating
+            self.shrinkage = shrinkage
+        else:
+            print('Unconstrained Execution')
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:50]: logger

File: src/torch_constrained/constrained_optimizer.py [hunk 10]
```diff
@@ -1,359 +1,289 @@
...
-
-        if not self.equality_multipliers and not self.inequality_multipliers:
-            self.init_dual_variables(eq_defect, inequality_defect, dtype=self.dual_dtype)
-
-        assert eq_defect is None or all([validate_defect(d, m) for d, m in zip(eq_defect, self.equality_multipliers)])
-        assert inequality_defect is None or all([d.shape == m.shape for d, m in zip(inequality_defect, self.inequality_multipliers)])
-
-        lagrangian = self.minmax_backward(loss, eq_defect, inequality_defect)
-
+            closure_dict = closure()
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:68]: returning dicts, is risky because it's hard to check their correctness at all time and specifying what should be inside and what should not be inside, also it's hard ensure the content doesn't change over time.
Jax solves this problem with namedtuples, I think that's a good tradeoff, additionally closure.equality is more readable than closure['equality']

File: src/torch_constrained/constrained_optimizer.py [hunk 11]
```diff
@@ -1,359 +1,289 @@
...
+
+            # JGP TODO: Deactivated this. What is this supposed to do?
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:70]: TODO

File: src/torch_constrained/constrained_optimizer.py [hunk 12]
```diff
@@ -1,359 +1,289 @@
...
+            # if self.shrinkage is not None and closure_dict['eq_defect']:
+            #     closure_dict['eq_defect'] = [self.shrinkage(e) for e in closure_dict['eq_defect']]
+
+            return closure_dict
+
+        closure_dict = closure_with_shrinkage()
+        loss, eq_defect, ineq_defect = [closure_dict[_] for _ in ['loss', 'eq_defect', 'ineq_defect']]
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:77]: this would become: loss, eq_defect, ineq_defect = closure_dict

File: src/torch_constrained/constrained_optimizer.py [hunk 13]
```diff
@@ -1,359 +1,289 @@
...
+
+            # Ensure multiplier shapes match those of the defects
+            assert eq_defect is None or all([validate_defect(d, m) for d, m in zip(eq_defect, self.eq_multipliers)])
+            assert ineq_defect is None or all([d.shape == m.shape for d, m in zip(ineq_defect, self.ineq_multipliers)])
+
+        # Compute lagrangian value based on current loss and values of multipliers
+        lagrangian = self.lagrangian_backward(loss, eq_defect, ineq_defect)
+        closure_dict['lagrangian'] = lagrangian
+        
+        # JGP TODO: Why was this being applied on the object loss?
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:92]: TODO

File: src/torch_constrained/constrained_optimizer.py [hunk 14]
```diff
@@ -1,359 +1,289 @@
...
-        if not self.alternating and hasattr(self.dual_optimizer, "extrapolation"):
+        if self.is_constrained and not self.alternating and hasattr(self.dual_optimizer, "extrapolation"):
             self.dual_optimizer.extrapolation(loss)
             should_back_prop = True
 
         if should_back_prop:
-            loss_, eq_defect_, inequality_defect_ = closure_with_shrinkage()
-            lagrangian_ = self.minmax_backward(loss_, eq_defect_, inequality_defect_)
+            closure_dict_ = closure_fn()
+            in_tuple = (closure_dict_[_] for _ in ['loss', 'eq_defect', 'ineq_defect'])
```
Comment (manuel-delverme) [inline src/torch_constrained/constrained_optimizer.py:112]: this is awkward, you can just skip the tuple that later gets unpacked

**Review by manuel-delverme — CHANGES_REQUESTED** [review]
I'm half way of the first file, but due to the amount of line changes it will have to happen slowly step by step like this.
